dashboard/app/models/concerns/levels/levels_within_levels.rb

```diff
@@ -0,0 +1,135 @@
+# The LevelsWithinLevels module encapsulates all functionality related to the
+# various ways we have of combining levels, including:
+#
+# - BubbleChoice
+# - Contained Levels
+# - LevelGroup
+# - Project Templates
+module Levels
+  module LevelsWithinLevels
+    extend ActiveSupport::Concern
+
+    included do
+      # Because this parent-child relationship comes in different flavors, we
+      # also want to provide some scopes to make interacting with subsets of
+      # relationships easier.
+      module ByKindExtension
+        ParentLevelsChildLevel::VALID_KINDS.each do |kind|
+          define_method(kind) do
+            where(parent_levels_child_levels: {kind: kind})
+          end
+        end
+      end
+
+      # We store parent-child relationships in a self-referential join table.
+      # In order to define a has_many/through relationship in both directions,
+      # we must define two separate associations to the same join table.
+      has_many :levels_parent_levels,
+        class_name: 'ParentLevelsChildLevel',
+        foreign_key: :child_level_id
+      has_many :parent_levels,
+        -> {extending ByKindExtension},
+        through: :levels_parent_levels,
+        inverse_of: :child_levels
+
+      has_many :levels_child_levels,
+        class_name: 'ParentLevelsChildLevel',
+        foreign_key: :parent_level_id
+      has_many :child_levels,
+        -> {extending ByKindExtension},
+        through: :levels_child_levels,
+        inverse_of: :parent_levels
+
+      before_validation :sanitize_contained_level_names
+      after_save :setup_contained_levels
+    end
+
+    # Returns all child levels of this level, which could include contained
+    # levels, project template levels, BubbleChoice sublevels, or LevelGroup
+    # sublevels.  This method may be overridden by subclasses.
+    def all_child_levels
+      (contained_levels + [project_template_level] - [self]).compact
+    end
+
+    # we must search recursively for child levels, because some bubble choice
+    # sublevels have project template levels.
+    def all_descendant_levels
+      my_child_levels = all_child_levels
+      child_descendant_levels = my_child_levels.
+        map(&:all_descendant_levels).flatten
+      my_child_levels + child_descendant_levels
+    end
+
+    # Helper method for retrieving contained levels; primarily exists to
+    # provide caching
+    def contained_levels
+      return [] if contained_level_names.blank?
+      return child_levels.contained unless Script.should_cache?
+      cache_key = "LevelsWithinLevels/contained/#{contained_level_names&.join('/')}"
+      Rails.cache.fetch(cache_key) do
+        levels_child_levels.map do |lcl|
+          Script.cache_find_level(lcl.child_level_id)
+        end
+      end
+    end
```

Comment:
I'm not at all sure how I feel about the implementation of this method; right now, it feels like a weird blend of the "old" approach (store relationship information in the `contained_level_names` field on the parent object, and directly "query" the script cache to retrieve the individual child objects) and the "new" approach (store relationship information in a many-to-many table, and query that table to get the children).

I think the specific problem here is that our solution to reduce unnecessary queries in production is optimized for the old approach, so in order for the new approach to avoid introducing a bunch of extra queries it needs to be so similar to the old approach that we're no longer really getting any benefits from the new approach.

Reply:
Does the regression in query counts affect only levels with contained levels, or all levels?

My initial thought is to focus on HOC scale scenarios when deciding whether query count regressions are ok. if this only adds query counts when viewing a level that has a contained level, then I would consider that to be acceptable, since we don't have many (any?) of those in HOC scripts.

More generally, I would love to prioritize getting onto the new system over exact query parity with the old implementation. the goal of this work is to reduce complexity and enable new features (such as the ability to discover which scripts and parent levels depend on a child level), and it seems ok to me for new functionality to come at a slight increase in query costs.

Does this help at all?

Reply:
That does help, thanks!

I'll put together a branch off this branch to highlight the required query count changes; I think it'll help to have specifics to look at.

Reply:
Here's that branch: https://github.com/code-dot-org/code-dot-org/pull/41588/files#r676830083

Turns out with the right caching in place, we don't actually add all that much with a simpler approach.

Reply:
Thanks @Hamms ! Just checking if you are ready for me to take another look at this?

Reply:
probably; I _think_ that the current test failure is just drone being problematic, but I'd like to get that green checkmark just in case. I'll re-request a review once it's fully ready

Reply:
Well, I still don't have the green checkmark but I've confirmed that the UI test failures are just flaky. Please go ahead and take a look when you can!

Reply:
After looking through some metrics-tracking tools with Suresh's help, I'm feeling quite comfortable that this change will not noticeably impact performance. I do plan to keep an eye on some things post-deploy, but I think for now we're ready to move forward with the new and improved implementation!